python-lib/dku_stress_test_center/stress_test_center.py

```diff
@@ -1,166 +1,177 @@
-    stress_test_indicator = perturbed_df[DkuStressTestCenterConstants.STRESS_TEST_TYPE]
-    valid_stress_ids = set(stress_test_indicator) & set(DkuStressTestCenterConstants.PERTURBATION_BASED_STRESS_TYPES)
-    valid_stress_ids |= set([DkuStressTestCenterConstants.CLEAN])
-
-    true_class_confidence_df = pd.DataFrame({
-        DkuStressTestCenterConstants.STRESS_TEST_TYPE: stress_test_indicator,
-        DkuStressTestCenterConstants.DKU_ROW_ID: perturbed_df[DkuStressTestCenterConstants.DKU_ROW_ID],
-        DkuStressTestCenterConstants.CONFIDENCE: y_true_class_confidence
-    })
-
-    true_class_confidence_df = true_class_confidence_df[true_class_confidence_df[DkuStressTestCenterConstants.STRESS_TEST_TYPE].isin(valid_stress_ids)]
-
-    # critical samples evaluation sorted by std of uncertainty
-
-    std_confidence_df = true_class_confidence_df.groupby([DkuStressTestCenterConstants.DKU_ROW_ID]).std()
-    uncertainty = np.round(std_confidence_df[DkuStressTestCenterConstants.CONFIDENCE].dropna(), 3)
-
-    critical_samples_df = pd.DataFrame({DkuStressTestCenterConstants.UNCERTAINTY: uncertainty})
-
-    if critical_samples_df.empty:
-        return pd.DataFrame(), []
-
-    critical_samples_df = critical_samples_df.sort_values(by=DkuStressTestCenterConstants.UNCERTAINTY,
-                                                          ascending=False)
+class StressTestGenerator(object):
+    def __init__(self, random_state=65537):
+        self._random_state = random_state
 
-    critical_samples_df.reset_index(level=0, inplace=True)
+        self.tests = None
+        self.model_accessor = None
+        self._sampling_proportion = None
+        self._clean_df = None
 
-    clean_df_with_id = perturbed_df.loc[perturbed_df[DkuStressTestCenterConstants.STRESS_TEST_TYPE] == DkuStressTestCenterConstants.CLEAN]
-    critical_samples_df = critical_samples_df.merge(clean_df_with_id, on=DkuStressTestCenterConstants.DKU_ROW_ID, how='left').head(top_k_samples)
-    return (critical_samples_df.drop([DkuStressTestCenterConstants.DKU_ROW_ID, DkuStressTestCenterConstants.STRESS_TEST_TYPE, DkuStressTestCenterConstants.UNCERTAINTY], axis=1),
-        critical_samples_df[DkuStressTestCenterConstants.UNCERTAINTY].tolist())
+    def set_config(self, config: dict):
+        self._sampling_proportion = config["samples"]
+        tests_config = config["perturbations"]
+        self.tests = defaultdict(list)
+        for test_name, test_config in tests_config.items():
+            test = StressTest.create(test_name, test_config)
+            self.tests[test.TEST_TYPE].append(test)
+
+    def compute_test_metrics(self, test: StressTest):
+        target = self.model_accessor.get_target_variable()
+        df = self.model_accessor.get_original_test_df(sample_fraction=self._sampling_proportion,
+                                                      random_state=self._random_state)
+        perturbed_df = test.perturb_df(df, target)
```

Comment:
For each test, we repeat the sampling step, while we're getting always the same data (thx to the random state). Wouldn't it be better to leave this part out of the `compute_stress_metrics` function ?

Reply:
But I don't have a strong opinion, just wondering if it takes longer without need.

Reply:
Indeed, I suppose it is a memory vs time trade off here, the alternative is to copy eveyrtime the df. Should be slightly faster, we can do it like this

Reply:
Done in 39cd1f1